### PR TITLE
Add hive table property 'retention_days'

### PR DESCRIPTION
**Pull request by 123subhadeep**

### CODE REVIEW

**Comment by electrum** [comment]
Wrong commit message

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 1]
```diff
@@ -963,6 +970,7 @@ private void rollbackPartitionUpdates(List<PartitionUpdate> partitionUpdates, St
 
     /**
      * Attempts to remove the file or empty directory.
+     *
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: These should not change, at least in this commit

**Comment by 123subhadeep** [comment]
Updated commit message.

**Comment by martint** [comment]
Please, add release notes.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java [hunk 2]
```diff
@@ -55,7 +58,8 @@ public HiveTableProperties(TypeManager typeManager, HiveClientConfig config)
                         false,
                         value -> ImmutableList.copyOf(((List<String>) value).stream()
                                 .map(name -> name.toLowerCase(ENGLISH))
-                                .collect(Collectors.toList()))));
+                                .collect(Collectors.toList()))),
+                integerSessionProperty(RETENTION_PROPERTY, "Table retention days", 0, false));
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: Is the default retention really 0 days?

Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: Should this be validated to prevent negative numbers?

**Comment by 123subhadeep** [comment]
In case of prism, the default retention value seems to be 0.
hive:bi> desc formatted temp_deep_30;
OK
# col_name            	data_type           	comment

col                 	string              	from deserializer

# Detailed Table Information
Database:           	bi
Owner:              	deep
CreateTime:         	Mon Nov 23 15:56:00 PST 2015
LastAccessTime:     	UNKNOWN
Protect Mode:       	None
Retention:          	0
Location:           	hdfs://dw-prn2-dfsgold.VOL5:9000/namespace/bi/warehouse/temp_deep_30
Table Type:         	MANAGED_TABLE
Table Parameters:
	RETENTION           	30
	comment             	Created by Presto
	transient_lastDdlTime	1448322960

# Storage Information
SerDe Library:      	org.apache.hadoop.hive.serde2.columnar.LazyBinaryColumnarSerDe
InputFormat:        	org.apache.hadoop.hive.ql.io.RCFileInputFormat
OutputFormat:       	org.apache.hadoop.hive.ql.io.RCFileOutputFormat
Compressed:         	No
Num Buckets:        	0
Bucket Columns:     	[]
Sort Columns:       	[]
Statistics stale:   	Yes
Time taken: 1.584 seconds, Fetched: 27 row(s)

**Comment by 123subhadeep** [comment]
What default value should I specify ?
The default retention seems to be 30 days in prism.

**Comment by 123subhadeep** [comment]
Added release notes
Added validation to accept only positive values for retention

**Comment by martint** [comment]
If the property is missing, the retention should be set to infinity.

**Comment by 123subhadeep** [comment]
So should I set the default value to Integer.MAX_VALUE ?

**Comment by martint** [comment]
If you don't set it at all, i think it will do just what we want.

**Comment by 123subhadeep** [comment]
Did the same.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 3]
```diff
@@ -394,7 +395,8 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
         List<String> partitionedBy = getPartitionedBy(tableMetadata.getProperties());
         List<HiveColumnHandle> columnHandles = getColumnHandles(connectorId, tableMetadata, ImmutableSet.copyOf(partitionedBy));
         HiveStorageFormat hiveStorageFormat = getHiveStorageFormat(tableMetadata.getProperties());
-        createTable(session, schemaName, tableName, tableMetadata.getOwner(), columnHandles, hiveStorageFormat, partitionedBy);
+        Optional<Integer> retentionDays = getRetentionDays(tableMetadata.getProperties());
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Use `OptionalInt`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveOutputTableHandle.java [hunk 4]
```diff
@@ -52,6 +55,7 @@ public HiveOutputTableHandle(
 
         this.partitionedBy = ImmutableList.copyOf(requireNonNull(partitionedBy, "partitionedBy is null"));
         this.tableOwner = requireNonNull(tableOwner, "tableOwner is null");
+        this.retentionDays = retentionDays;
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveOutputTableHandle.java]: Check for null (the `Optional` reference should never be null, even though what it contains might be)

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java [hunk 5]
```diff
@@ -72,4 +78,16 @@ public static HiveStorageFormat getHiveStorageFormat(Map<String, Object> tablePr
     {
         return (List<String>) tableProperties.get(PARTITIONED_BY_PROPERTY);
     }
+
+    public static Optional<Integer> getRetentionDays(Map<String, Object> tableProperties)
+    {
+        if (tableProperties.containsKey(RETENTION_PROPERTY)) {
+            int retentionDays = (Integer) tableProperties.get(RETENTION_PROPERTY);
+            if (retentionDays != DEFAULT_RETENTION_DAYS) {
+                checkState(retentionDays > 0, String.format("%s must be positive", RETENTION_PROPERTY));
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: `checkState()` already takes a format string

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: This message doesn't match the check. Make the check `>= 0`

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: And make the message "cannot be negative"

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: Make the message "must be greater than zero"

**Comment by electrum** [comment]
Minor comments, otherwise looks good

**Comment by martint** [comment]
Applied changes and merged.

**Pull request closed without merging**